### PR TITLE
check if new function name is in lowercase

### LINKED ISSUE #374: build error while image name is uppercase
`faas-cli build` command failed when image name has a uppercase tag as
For example: 
`faas-cli build `
the internal `docker build -t <image_name> .` fails

## Expected Behaviour
```bash
$ faas-cli new myFunc --lang go
$ faas-cli build --yaml myFunc.yml
```
A image should be created have a image name as of my function name if not specified

## Current Behaviour
The `faas-cli build` command fails with below error: 
```bash
invalid argument "myFunc" for "-t, --tag" flag: invalid reference format: repository name must be lowercase
See 'docker build --help'.
2018/04/10 13:08:17 ERROR - Could not execute command: [docker build -t myFunc .]
```

## Possible Solution
This bug is introduced as docker build doesn't allow repository name to have uppercase
```bash
invalid argument "myFunc" for "-t, --tag" flag: invalid reference format: repository name must be lowercase
```
The bug can be omitted by always converting image name to lowercase before using any docker command locally or requesting to the gateway, even if they are in uppercase in `stack.yml` or in `--build` flag.

Otherwise lower case image name can be enforced by checking if `image` values and return error in case it is `Uppercase`

## Steps to Reproduce (for bugs)
```bash
// STEP 1
$ faas-cli new myFunc --lang go
// STEP 2
$ faas-cli build --yaml myFunc.yml
```

## Context
This issue currently forces the user to use lower case Image Name 
and when generated a new function provide the Image Name seperately

## Your Environment
* FaaS-CLI version ( Full output from: `faas-cli version` ):
```bash
Commit: ae7390005a2fe13873f2cb6fcfa2d830dad4a40b
Version: 0.6.5
```

* Docker version ( Full output from: `docker version` ):
```bash
Client:
 Version:	18.03.0-ce
 API version:	1.37
 Go version:	go1.9.4
 Git commit:	0520e24
 Built:	Wed Mar 21 23:10:09 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.03.0-ce
  API version:	1.37 (minimum version 1.12)
  Go version:	go1.9.4
  Git commit:	0520e24
  Built:	Wed Mar 21 23:08:36 2018
  OS/Arch:	linux/amd64
  Experimental:	false
```

* Are you using Docker Swarm (FaaS-swarm ) or Kubernetes (FaaS-netes)?
FaaS-swarm

* Operating System and version (e.g. Linux, Windows, MacOS):
Linux

* Link to your project or a code example to reproduce issue:

### PR DESCRIPTION
**Pull request by s8sg**
Signed-off-by: Swarvanu Sengupta <swarvanusg@gmail.com>

## Description
as docker build only support lowercase in repo name
name for new functions are always lowercase

## Motivation and Context
when user generate new function, a function name may contains `uppercase` i.e.
`myAwesomeFunc`, although as docker build doesn't support `uppercase` repo name
it fails. Described in details at: [issue-374](https://github.com/openfaas/faas-cli/issues/374)

So we only allow lowercase function name now

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
Added automated test case as:
```bash
     {
               title:       "test_Uppercase",
               funcName:    "test_Uppercase",
               funcLang:    "go",
               expectedMsg: IncludeUpperCase,
      }
```
Manually can be testes as: 
```
$ faas-cli new myAwesome_Func --lang go
Function name must be lowercase
```

## Types of changes

Fixes #374 
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Please see comments in the issue. I don’t think we will merge it as is but have suggested an alternative flow.

File: commands/new_function.go [hunk 1]
```diff
@@ -46,10 +47,38 @@ language or type in --list for a list of languages available.`,
...
+	}
+
+	if len(args) < 1 {
+		return fmt.Errorf("please provide a name for the function")
+	}
+
+	functionName = args[0]
+
+	if !containsLowerCase(functionName) {
+		return fmt.Errorf("Function name must be lowercase")
```
Comment (alexellis) [inline commands/new_function.go]: Hi, the errors should start with a lowercase letter too.

File: commands/new_function.go [hunk 2]
```diff
@@ -74,16 +103,6 @@ func runNewFunction(cmd *cobra.Command, args []string) error {
 		return nil
 	}
 
-	if len(args) < 1 {
```
Comment (alexellis) [inline commands/new_function.go]: What happened to these lines of code?

File: commands/new_function.go [hunk 3]
```diff
@@ -46,10 +47,38 @@ language or type in --list for a list of languages available.`,
 	RunE:    runNewFunction,
 }
 
+func containsLowerCase(funcname string) bool {
```
Comment (alexellis) [inline commands/new_function.go]: We could perhaps use a regex here?

What are the valid names of services on Kubernetes? They are in this regex:

https://github.com/openfaas/faas-netes/blob/04888f560506c452ed684c4d376ff08fbfe5daea/handlers/deploy.go#L39

Is this regex also compatible with the "image" field for Docker image names?

File: commands/new_function.go (hunk 2 above)
Comment (s8sg) [reply commands/new_function.go]: The validations are moved to [preRunNewFunction()](https://github.com/openfaas/faas-cli/pull/375/files/8df892ee565df83faac1726f7c0b2b793b82f229#diff-4e0fbf5ad1c657e2d3f8a554849f6f25R68)

**Comment by alexellis** [comment]
Derek close: via #385

**Pull request closed without merging**